File: hud/view/workout/workout_window.py
from PySide6.QtCore import Qt
from PySide6.QtWidgets import QGridLayout, QWidget

from hud.configuration.config import Config
from hud.model.model import Model
from hud.view import DeviceController
from hud.view.app_state import AppState
from hud.view.primitives.draggable_window import AppWindow
from hud.view.primitives.labels import ButtonLabel
from hud.view.primitives.theme_switch import with_switchable_theme
from hud.view.workout.metrics_panel import MetricsPanel


@with_switchable_theme
class WorkoutWindow(AppWindow):
    def __init__(
            self,
            app_config: Config,
            controller: DeviceController,
            model: Model,
            parent=None,
    ):
        super().__init__(app_config, parent)
        self.model = model
        self.controller = controller

        self.layout = None
        self.centralWidget = None

        self.heart_rate_monitor_panel = MetricsPanel(
            model=model,
            controller=controller,
            text="Heart Rate",
            app_config=self.app_config,
        )
        self.heart_rate_monitor_panel.bind_to_model(model.hrm_notifications)

        self.cadence_sensor_panel = MetricsPanel(
            model=model,
            controller=controller,
            text="Cadence",
            app_config=self.app_config,
        )
        self.cadence_sensor_panel.bind_to_model(model.cad_notifications)

        self.power_meter_panel = MetricsPanel(
            model=model,
            controller=controller,
            text="Instant Power",
            app_config=self.app_config,
        )
        self.power_meter_panel.bind_to_model(model.pwr_notifications)

        self.speed_sensor_panel = MetricsPanel(
            model=model,
            controller=controller,
            text="Speed",
            app_config=self.app_config,
        )
        self.speed_sensor_panel.bind_to_model(model.spd_notifications)

        self.finishLabel = ButtonLabel(
            text="Finish",
            theme=self.app_config.hud_layout.theme,
        )
        self.finishLabel.clicked.connect(lambda: self.next.emit(AppState.WORKOUT_FINISHED))

        self.layout = QGridLayout()
        self.centralWidget = QWidget(self)
        self.centralWidget.setLayout(self.layout)
        self.setCentralWidget(self.centralWidget)

        self.layout.addWidget(self.heart_rate_monitor_panel, 0, 0)
        self.layout.addWidget(self.cadence_sensor_panel, 1, 0)
        self.layout.addWidget(self.power_meter_panel, 2, 0)
        self.layout.addWidget(self.speed_sensor_panel, 3, 0)
        self.layout.addWidget(self.finishLabel, 4, 0, 1, 1, Qt.AlignmentFlag.AlignCenter)

File: src/WorkoutWindow.cpp
#include "SelectDevicePanel.h"
#include "StyleSheets.h"
#include "WorkoutWindow.h"

#include <iostream>

#include <Labels.h>

#include <QGridLayout>

#include "Constants.h"
#include "Events.h"


WorkoutWindow::WorkoutWindow(const std::shared_ptr<ControllerHandler> &handler, QWidget *parent): AppWindow(
    handler, parent) {
    eventHandlers.insert({
        getMeasurementReceivedType(), [this](QEvent *event) {
            const auto data = dynamic_cast<MeasurementReceivedEvent *>(event);
            std::cout << "WorkoutWindow::MeasurementReceivedEvent" << std::endl;
            measurementsReceived(data->getData());
        }
    });

    heartRateMonitorPanel = new MetricsPanel(Data::HEART_RATE, this);
    cadenceSensorPanel = new MetricsPanel(Data::CADENCE, this);
    powerMeterPanel = new MetricsPanel(Data::SPEED, this);
    speedSensorPanel = new MetricsPanel(Data::POWER, this);

    const auto finishLabel = new ButtonLabel(Constants::Buttons::FINISH, this);
    connect(finishLabel, &ButtonLabel::clicked, this, &WorkoutWindow::next);

    const auto layout = new QGridLayout(this);
    const auto centralWidget = new QWidget(this);

    centralWidget->setLayout(layout);
    setCentralWidget(centralWidget);

    layout->addWidget(heartRateMonitorPanel, 0, 0);
    layout->addWidget(cadenceSensorPanel, 1, 0);
    layout->addWidget(powerMeterPanel, 2, 0);
    layout->addWidget(speedSensorPanel, 3, 0);
    layout->addWidget(finishLabel, 4, 0, 1, 1, Qt::AlignmentFlag::AlignCenter);
}

void WorkoutWindow::measurementsReceived(const WorkoutData &data) {
    heartRateMonitorPanel->measurementsReceived(data);
    cadenceSensorPanel->measurementsReceived(data);
    speedSensorPanel->measurementsReceived(data);
    powerMeterPanel->measurementsReceived(data);
}

void WorkoutWindow::next() {
    std::cout << "WorkoutWindow::next" << std::endl;
    controllerHandler->next(Constants::Commands::QUIT);
}
